plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,654 @@
+			break;
+		case 1:
+			g_prefix_error (error,
+				  "NVM Auth write operation error");
+			break;
+		case 2:
+			g_prefix_error (error,
+				  "NVM Auth operation status: Retry NVM write");
+			break;
+		default:
+			g_prefix_error (error,
+				  "NVM Auth operation status unknown");
+			break;
+	}
+	/* FIXME pass op_status to caller in some way */
+	return FALSE;
+}
+
+static gboolean
+fu_dell_dock_usb4_activate (FuDevice *device, GError **error)
+{
+	return fu_dell_dock_usb4_hub_nvm_authenticate (device, error);
+}
+
+static const gchar*
+fu_dell_dock_usb4_show_nvm_info (FuDevice *device, GError **error)
+{
+	gchar *nvm_version;
+	guchar buf[64];
+	guint len;
+
+	len = 64;
+	if (!fu_dell_dock_usb4_hub_nvm_read (device, buf, len, 0, error)) {
+		g_prefix_error (error, "NVM Version Read Failed");
+		return NULL;
+	}
+	nvm_version = g_strdup_printf ("%02x.%02x", buf[10], buf[9]);
+	return nvm_version;
+}
+
+static gboolean
+fu_dell_dock_usb4_write_fw (FuDevice *device, FuFirmware *firmware,
+			    FwupdInstallFlags flags, GError **error)
+{
+	FuDellDockUsb4 *self = FU_DELL_DOCK_USB4 (device);
+	guint32 header_offset = 0;
+	gsize image_size = 0;
+	const guint8 *buffer;
+	g_autoptr(GTimer) timer = g_timer_new ();
+	g_autofree gchar *dynamic_version    = NULL;
+	g_autofree gchar *dynamic_vendor_id  = NULL;
+	g_autofree gchar *dynamic_product_id = NULL;
+	g_autoptr(GBytes) fw = NULL;
+
+	g_return_val_if_fail (device != NULL, FALSE);
+	g_return_val_if_fail (FU_IS_FIRMWARE (firmware), FALSE);
+
+	/* get default image */
+	fw = fu_firmware_get_bytes (firmware, error);
+	if (fw == NULL)
+		return FALSE;
+	buffer = g_bytes_get_data (fw, &image_size);
+	g_debug ("Total Image size: %" G_GSIZE_FORMAT, image_size);
+
+	/* get image version */
+	dynamic_version = g_strdup_printf ("%02x.%02x",
+					   buffer[self->blob_major_offset],
+					   buffer[self->blob_minor_offset]);
+	g_debug ("writing Thunderbolt firmware version %s", dynamic_version);
+
+	/* get image vendor-id */
+	dynamic_vendor_id = g_strdup_printf ("0x%02x%02x",
+					     buffer[self->vid_offset+1],
+					     buffer[self->vid_offset]);
+	g_debug ("writing Thunderbolt vendor-id %s", dynamic_vendor_id);
+
+	/* get image product-id */
+	dynamic_product_id = g_strdup_printf ("0x%02x%02x",
+					       buffer[self->pid_offset+1],
+					       buffer[self->pid_offset]);
+	g_debug ("writing Thunderbolt product-id %s", dynamic_product_id);
+
+	/* check header size */
+	memcpy (&header_offset, buffer, sizeof (guint32));
+	g_debug ("Header size: %" G_GUINT32_FORMAT, header_offset);
+	if (header_offset > image_size) {
+		g_set_error (error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+			     "Image header is too big: %" G_GUINT32_FORMAT,
+			     header_offset);
+		return FALSE;
+	}
+	buffer += header_offset;
+	image_size -= header_offset;
+
+	/* firmware install */
+	if (!fu_dell_dock_usb4_hub_nvm_write(device, buffer, image_size, 0,
+					     error))
+		return FALSE;
+
```

Comment:
don't you need to set a needs activation flag here?

Reply:
I'm testing the `activation` by running `fwupdtool install` with `USB4` device only. With the addition of `FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION` here right after writing the firmware the `engine` somehow doesn't exercise `vfunc->activate` at all, it's weird.

With `fwupdtool activate` command the `needs_activation` isn't set yet therefore [quit](https://github.com/fwupd/fwupd/blob/master/src/fu-tool.c#L1702) early.
```
if (!fu_device_has_flag (device, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION))
     continue;
```

How to launch `activate` properly?

Reply:
>I'm testing the activation by running fwupdtool install with USB4 device only. With the addition of FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION here right after writing the firmware the engine somehow doesn't exercise vfunc->activate at all, it's weird.

That's expected behavior.  If you have that flag set, engine stops after write.  Then you need a separate activate call or disconnect action to finish.

>With fwupdtool activate command the needs_activation isn't set yet therefore quit early.

If this splits to a separate plugin (as my fundamental question above proposed) you'll need to modify this to cover both plugins:
```
		if (g_strcmp0 (fu_device_get_plugin (dev), "thunderbolt") == 0 
```
https://github.com/fwupd/fwupd/blob/master/plugins/dell-dock/fu-plugin-dell-dock.c#L249

If it stays in same plugin you'll need to call:
```
				fu_dell_dock_ec_tbt_passive (parent);
```
https://github.com/fwupd/fwupd/blob/master/plugins/dell-dock/fu-plugin-dell-dock.c#L253